### PR TITLE
combine .sample() and .sample_n()

### PR DESCRIPTION
**Pull request by alicanb**
@apaszke we (mostly @fritzo, I just had some free time to code 😃 ) thought if we combine `.sample_n` and `sample`, we can make `.rsample(n)` in #3886 and avoid implementing a separate `.rsample_n`.

### CODE REVIEW

File: torch/distributions.py [hunk 1]
```diff
@@ -53,7 +55,7 @@ class Distribution(object):
     Distribution is the abstract base class for probability distributions.
     """
 
-    def sample(self):
+    def sample(self, n):
```
Comment (fritzo) [inline torch/distributions.py]: `def sample(self, n=None)`

**Review by fritzo — COMMENTED** [review]
Thanks for doing this! Consider also replacing `sample_n` with `sample` everwhere in `tests/test_distributions.py`.

**Comment by alicanb** [comment]
I wanted to make sure it's fully backwards compatible, we will write new tests when we introduce `.sample(sample_shape)` anyway.

**Review by fritzo — APPROVED** [review]
LGTM (from a non-owner)

File: torch/distributions.py [hunk 2]
```diff
@@ -53,9 +55,9 @@ class Distribution(object):
     Distribution is the abstract base class for probability distributions.
     """
 
-    def sample(self):
+    def sample(self, n=None):
```
Comment (apaszke) [inline torch/distributions.py]: wasn't the plan to accept shape tuples instead of a single number?

**Comment by fritzo** [comment]
I believe this PR is a small step in that direction, and helps simplify
concurrent PRs. In the future sample_shape should default to ().

**Comment by apaszke** [comment]
I’d rather limit the number of intermediate changes to the interface to not confuse people who are actually using master. Can we just make the change in one go? It shouldn’t be too complicated. But I can merge this as is if you expect it to be fixed quickly

**Comment by fritzo** [comment]
Ok, if we want to avoid introducing a temporary`rsample_n()` method in #3886 , I think the shortest path would be to change this PR to a limited version of the final form of `.sample()`:
```py
def sample(sample_shape=()):
    if len(sample_shape) == 0:
        n = None 
        # use logic from .sample()
    elif len(sample_shape) == 1:
        n = sample_shape[0] 
        # use logic from .sample_n(n)
    else:
        raise NotImplementedError
```
What do you think?

**Comment by apaszke** [comment]
That sounds good to me. Implementation can change a lot, but let’s minimize the periods with temporary interfaces

File: torch/distributions.py [hunk 3]
```diff
@@ -65,7 +66,8 @@ def sample_n(self, n):
         Generates n samples or n batches of samples if the distribution parameters
         are batched.
         """
-        raise NotImplementedError
+        warnings.warn('sample_n will be deprecated. Use .sample(n) instead', PendingDeprecationWarning)
```
Comment (apaszke) [inline torch/distributions.py]: n will be need to wrapped in a tuple. The code that this warning suggests doesn’t work

File: torch/distributions.py [hunk 4]
```diff
@@ -146,14 +150,16 @@ def __init__(self, probs):
...
-
-    def sample_n(self, n):
-        if n == 1:
-            return self.sample().expand(1, 1)
+    def sample(self, sample_shape=()):
+        if len(sample_shape) == 0:
+            return torch.multinomial(self.probs, 1, True).squeeze(-1)
+        elif len(sample_shape) == 1:
+        	if sample_shape[0] == 1:
+        		return self.sample().expand(1, 1)
```
Comment (apaszke) [inline torch/distributions.py]: There’s something wrong with indentation here. Tabs?

**Comment by fritzo** [comment]
Close in favor of https://github.com/pytorch/pytorch/pull/4142 ?

**Pull request closed without merging**